src/package-request.js

```diff
@@ -281,7 +281,8 @@ export default class PackageRequest {
       promises.push(this.resolver.find({
         pattern: depPattern,
         registry: remote.registry,
-        optional: false,
+        // dependencies of optional dependencies should themselves be optional
+        optional: this.optional,
```

Comment:
What happens if a dependency is optional in one subtree and is not optional in another:
```
A -> (optional) B -> C
D -> C
```

Would it mark C as optional even though D requires it as a regular dependency?
Could you add a test for that to make sure we don't break it?